packages/eslint-plugin/src/rules/prefer-destructuring.ts

```diff
@@ -0,0 +1,239 @@
+  ],
+  create(context, [enabledTypes, options1 = {}]) {
+    const {
+      enforceForRenamedProperties = false,
+      enforceForDeclarationWithTypeAnnotation = false,
+    } = options1;
+    const { program, esTreeNodeToTSNodeMap } = getParserServices(context);
+    const typeChecker = program.getTypeChecker();
+    const baseRules = baseRule.create(context);
+    let baseRulesWithoutFixCache: typeof baseRules | null = null;
+
+    return {
+      VariableDeclarator(node): void {
+        performCheck(node.id, node.init, node);
+      },
+      AssignmentExpression(node): void {
+        if (node.operator !== '=') {
+          return;
+        }
+        performCheck(node.left, node.right, node);
+      },
+    };
+
+    function performCheck(
+      leftNode: TSESTree.BindingName | TSESTree.Expression,
+      rightNode: TSESTree.Expression | null,
+      reportNode: TSESTree.VariableDeclarator | TSESTree.AssignmentExpression,
+    ): void {
+      const rules =
+        leftNode.type === AST_NODE_TYPES.Identifier &&
+        leftNode.typeAnnotation === undefined
+          ? baseRules
+          : baseRulesWithoutFix();
+      if (
+        'typeAnnotation' in leftNode &&
+        leftNode.typeAnnotation !== undefined &&
+        !enforceForDeclarationWithTypeAnnotation
+      ) {
+        return;
+      }
+
+      if (
+        rightNode != null &&
+        isArrayLiteralIntegerIndexAccess(rightNode) &&
+        rightNode.object.type !== AST_NODE_TYPES.Super
+      ) {
+        const tsObj = esTreeNodeToTSNodeMap.get(rightNode.object);
+        const objType = typeChecker.getTypeAtLocation(tsObj);
+        if (!isTypeAnyOrIterableType(objType, typeChecker)) {
+          if (
+            !enforceForRenamedProperties ||
+            !getNormalizedEnabledType(reportNode.type, 'object')
+          ) {
+            return;
+          }
+          context.report({
+            node: reportNode,
+            messageId: 'preferDestructuring',
+            data: { type: 'object' },
+          });
+          return;
+        }
+      }
+
+      if (reportNode.type === AST_NODE_TYPES.AssignmentExpression) {
+        rules.AssignmentExpression(reportNode);
+      } else {
+        rules.VariableDeclarator(reportNode);
+      }
+    }
+
+    function getNormalizedEnabledType(
+      nodeType:
+        | AST_NODE_TYPES.VariableDeclarator
+        | AST_NODE_TYPES.AssignmentExpression,
+      destructuringType: 'array' | 'object',
+    ): boolean | undefined {
+      if ('object' in enabledTypes || 'array' in enabledTypes) {
+        return enabledTypes[destructuringType];
+      }
+      return enabledTypes[nodeType as keyof typeof enabledTypes][
+        destructuringType as keyof (typeof enabledTypes)[keyof typeof enabledTypes]
+      ];
+    }
+
+    function baseRulesWithoutFix(): ReturnType<typeof baseRule.create> {
+      if (baseRulesWithoutFixCache == null) {
+        baseRulesWithoutFixCache = baseRule.create(noFixContext(context));
+      }
```

Comment:
[Nitpick] Just a little cleaner here:

```suggestion
      baseRulesWithoutFixCache ??= baseRule.create(noFixContext(context));
```

I'm actually surprised we don't have https://eslint.org/docs/latest/rules/logical-assignment-operators enabled... #7461